source/fuzz/fact_manager.cpp

```diff
@@ -1225,30 +1228,46 @@ bool FactManager::DataSynonymAndIdEquationFacts::
   // vectors that differ only in signedness.
 
   // Get both types.
-  const opt::analysis::Type* type_1 =
-      context->get_type_mgr()->GetType(end_type_id_1);
-  const opt::analysis::Type* type_2 =
-      context->get_type_mgr()->GetType(end_type_id_2);
-
-  // If the first type is a vector, check that the second type is a vector of
-  // the same width, and drill down to the vector element types.
-  if (type_1->AsVector()) {
-    if (!type_2->AsVector()) {
-      return false;
-    }
-    if (type_1->AsVector()->element_count() !=
-        type_2->AsVector()->element_count()) {
-      return false;
-    }
-    type_1 = type_1->AsVector()->element_type();
-    type_2 = type_2->AsVector()->element_type();
+  const auto* type_a = context->get_type_mgr()->GetType(end_type_id_1);
+  const auto* type_b = context->get_type_mgr()->GetType(end_type_id_2);
+  assert(type_a && type_b && "Data descriptors have invalid type(s)");
+
+  uint32_t element_count_a = 1;
+  uint32_t element_count_b = 1;
```

Comment:
I suggest you replace from here to the end of the function with:

```
auto get_bit_count_for_numeric_type = // your nice lambda function;
if ((type_a->AsFloat() || type_a->AsInt()) && (type_b->AsFloat() || type_b->AsInt())) {
  return get_bit_count_for_numeric_type(type_a) == get_bit_count_for_numeric_type(type_b);
}
if (type_a->AsVector() && type_b->AsVector()) {
  if (either element type is bool) {
    return false; // The case where both are bool is handled by the direct equality check earlier; you can't have multiple identical boolean vector types
  }
  return true if and only if element counts are equal and bit counts of element types are equal
}
return false;
```

Reply:
Fixed in 2955d7a.